# Sample quality control
We now know from reading [Removing the bad apples: A simple bioinformatic method to improve loci-recovery in de novo RADseq data for non-model organisms](https://besjournals.onlinelibrary.wiley.com/doi/full/10.1111/2041-210X.13562), that individual samples with low data can negatively affect our ability to call shared SNPs across our datasets. We will take a simple approach to removing low data samples, by running [fastqcR](https://rpkgs.datanovia.com/fastqcr/index.html) on our demultiplexed sample files.
### Step 1:
> 



### Next week:
> Homework for next week is to read the [Stacks manual](https://catchenlab.life.illinois.edu/stacks/manual/) to get as familiar as possible with the pipeline we will be implementing next week. Please spend as much time as you can familiarizing yourself with this pipeline, understanding it will be key to working with your RAD datasets!
