---
title: "Petroica RAD preliminary analyses"
author: "Devon DeRaad"
date: '2022-08-06'
output: html_document
---

# Start by making a splitstree
```{r}
library(vcfR)
library(SNPfiltR)
library(StAMPP)
library(adegenet)
library(ggplot2)

#read vcf
v<-read.vcfR("~/Desktop/petroica.preliminary/populations.snps.vcf")
#remove hopeless samples
missing_by_sample(v)
v<-missing_by_sample(v, cutoff = .7)
#remove hopeless SNPs
missing_by_snp(v)
v<-missing_by_snp(v, cutoff = .85)
#convert to genlight
gen<-vcfR2genlight(v)
#
gen@ind.names
gen@ind.names<-gsub("P_pusilla","pus", gen@ind.names)
gen@ind.names<-gsub("P_goodenovii","good", gen@ind.names)
pop(gen)<-gen@ind.names
sample.div <- stamppNeisD(gen, pop = FALSE)
#export for splitstree
#stamppPhylip(distance.mat=sample.div, file="~/Desktop/petroica.preliminary/petroica.85.splits.txt")
#plot the resulting splitstree
knitr::include_graphics("/Users/devder/Desktop/petroica.preliminary/splitstree.with.birds.png")
```

# plot pairwise Fst
```{r}
#calculate Fst and fixed differences
#make sampling df in same order as input vcf
samps<-data.frame(id=gen@ind.names,
                  species=as.factor(c("makira","fiji","fiji","fiji","fiji","fiji","fiji",
           "guad","guad","aus","aus","makira","makira","makira",
           "makira","kolo","kolo","kolo","kolo")))
#calc pairwise Fst
gen@ind.names
gen@pop<-samps$species
di.heat<-stamppFst(gen)
m<-di.heat$Fsts
#fill in upper triangle of matrix
m[upper.tri(m)] <- t(m)[upper.tri(m)]

#melt for plotting
heat <- reshape::melt(m)

#plot with labels
ggplot(data = heat, aes(x=X1, y=X2, fill=value)) + 
  geom_tile()+
  geom_text(data=heat,aes(label=round(value, 2)))+
  theme_minimal()+
  scale_fill_gradient2(low = "white", high = "red", space = "Lab", name="Fst") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.text.y = element_text(angle = 45, hjust = 1))

#identify the number of pairwise fixed differences between pops
mat<-extract.gt(v)
conv.mat<-mat
conv.mat[conv.mat == "0/0"]<-0
conv.mat[conv.mat == "0/1"]<-1
conv.mat[conv.mat == "1/1"]<-2
conv.mat<-as.data.frame(conv.mat)
#convert to numeric
for (i in 1:ncol(conv.mat)){
  conv.mat[,i]<-as.numeric(as.character(conv.mat[,i]))
}

#show colnames to verify you're subsetting correctly
colnames(conv.mat) == samps$id

#make vector to fill with fixed diff values
f<-c()

#write for loop to calc number of fixed diffs between each pop
for (i in 1:nrow(heat)){
  #calc af of pop1 and pop2
  pop1.af<-(rowSums(conv.mat[,samps$species == heat$X1[i]], na.rm=T)/(rowSums(is.na(conv.mat[,samps$species == heat$X1[i]]) == FALSE)))/2
  pop2.af<-(rowSums(conv.mat[,samps$species == heat$X2[i]], na.rm=T)/(rowSums(is.na(conv.mat[,samps$species == heat$X2[i]]) == FALSE)))/2
  #store number of fixed differences
  f[i]<-sum(is.na(abs(pop1.af - pop2.af)) == FALSE & abs(pop1.af - pop2.af) == 1) #find fixed SNPs and add to vector
}

#add number of fixed diffs to df
heat$fixed<-f

###code that will get you the vector needed to fix assignments of FST values versus fixed differences
#define n as the number of taxa used in your pairwise comparison
n<-5
i<-1 #begin incrementer (i) at 1
#while loop that will make the appropriate vector and store it in the variable 'x'
while (i < n){
  #the first set of numbers is simply 2:n
  if(i == 1){
    x<-c(2:n)
    i=i+1
  }
  #the second set of numbers is (2+n+1):(2*n) which we add to the existing vector
  if(i == 2){
    x<-c(x,(2+n+1):(2*n))
    i=i+1
  }
  #we then add (2+((i-1)*(n+1))):(i*n) to the vector, where i=3, and continue adding this vector to the growing vector until i = n-1
  if(i > 2){
    x<-c(x,(2+((i-1)*(n+1))):(i*n))
    i=i+1
  }
}

#fix the assignments using the vector stored as 'x'
heat$mixed<-heat$value
heat$mixed[x]<-heat$fixed[x]

#plot with labels
ggplot(data = heat, aes(x=X1, y=X2, fill=value)) + 
  geom_tile()+
  geom_text(data=heat,aes(label=round(mixed, 2)), size=4)+
  theme_minimal()+
  scale_fill_gradient2(low = "white", high = "red", space = "Lab", name="Fst") +
  theme(axis.text.x = element_text(angle = 45, vjust=.9, hjust = .9, size=12),
        axis.text.y = element_text(angle = 45, hjust = 1, size=12),
        axis.title.x = element_blank(), axis.title.y = element_blank())
```

# plot heterozygosity for each group
```{r}
#run this in bash, where retained.popmap.txt assigns samples to species, to get species level pi estimates
#and todi.singlesample.popmap.txt assigns each sample as a unique population to get per individual heterozygosity estimates
#mkdir fixed.pops
#mkdir single.sample
##Run populations and export population info
#/home/d669d153/work/stacks-2.41/populations -P ./stacks.prelim -M pops.popmap.txt -O ./fixed.pops
##Run populations and export population info
#/home/d669d153/work/stacks-2.41/populations -P ./stacks.prelim -M singlesample.popmap.txt -O ./single.sample

#open each file in excel and delete the "only variant positions" info (retaining "all positions variant and fixed") and remove leading hash's (#)

#read in each output file
#pops
pi.pops<-read.table("~/Desktop/petroica.preliminary/pops.populations.sumstats_summary.tsv", header=T, sep='\t')
#per sample
pi.sample<-read.table("~/Desktop/petroica.preliminary/sample.populations.sumstats_summary.tsv", header=T, sep='\t')

#add species to per sample df
pi.sample$species<-samps$species
#make tidy format pi vector
g<-c()
for(i in 1:nrow(pi.sample)){
  g[i]<-pi.pops$Pi[pi.pops$Pop.ID == pi.sample$species[i]]
}

plotting.df<-data.frame(sample=pi.sample$Pop.ID,
                        species=samps$species,
                        het=pi.sample$Obs_Het,
                        pi=g)
#plot heterozygosity violin plots
ggplot(plotting.df, aes(x=species, y=het)) + 
  geom_dotplot(binaxis='y', stackdir='center', dotsize = 1, alpha=.75)+
  theme_classic()+
  geom_point(plotting.df, mapping=aes(y=pi), pch=8, cex=3, col="red")+
  labs(x="",y="heterozygosity")+
  scale_y_continuous(sec.axis = sec_axis(trans = (~.*1), name="Pi"))
```
# admixture plots
```{r}
#distance thin vcf for only unlinked SNPs
#y<-distance_thin(v, min.distance = 1000)
#must make chromosome names non-numeric for plink
#y@fix[,1]<-paste("a", y@fix[,1], sep="")
#write thinned file to disk using vcfR
#vcfR::write.vcf(y, file="~/Desktop/petroica.preliminary/filtered.unlinked.85.vcf.gz")

#use this thinned file to execute ADMIXTURE on the cluster using this script:
##!/bin/sh
##
##SBATCH --job-name=admixture               # Job Name
##SBATCH --nodes=1             # 40 nodes
##SBATCH --ntasks-per-node=10               # 40 CPU allocation per Task
##SBATCH --partition=sixhour         # Name of the Slurm partition used
##SBATCH --chdir=/home/d669d153/scratch/petroica/admixture     # Set working d$
##SBATCH --mem-per-cpu=1gb            # memory requested
##SBATCH --time=360
#
##use plink to convert vcf directly to bed format:
#/home/d669d153/work/plink --vcf filtered.unlinked.85.vcf --double-id --allow-extra-chr --make-bed --out binary_fileset
##fix chromosome names
#cut -f2- binary_fileset.bim  > temp
#awk 'BEGIN{FS=OFS="\t"}{print value 1 OFS $0}' temp > binary_fileset.bim

##run admixture for a K of 1-10, using cross-validation, with 10 threads
#for K in 1 2 3 4 5 6 7 8 9 10; 
#do /home/d669d153/work/admixture_linux-1.3.0/admixture --cv -j10 binary_fileset.bed $K | tee log${K}.out;
#done
#
##Which K iteration is optimal according to ADMIXTURE ?
#grep -h CV log*.out > log.errors.txt

#read in ADMIXTURE results
#setwd to admixture directory run on the cluster
setwd("~/Desktop/petroica.preliminary/admixture/")

#read in log error values to determine optimal K
log<-read.table("log.errors.txt")[,c(3:4)]
#use double backslash to interpret the opening parentheses literally in the regular expression
log$V3<-gsub("\\(K=", "", log$V3)
log$V3<-gsub("):", "", log$V3)
#interpret K values as numerical
log$V3<-as.numeric(log$V3)
#rename columns
colnames(log)<-c("Kvalue","cross.validation.error")

#make plot showing the cross validation error across K values 1:10
ggplot(data=log, aes(x=Kvalue, y=cross.validation.error, group=1)) +
  geom_line(linetype = "dashed")+
  geom_point()+
  ylab("cross-validation error")+
  xlab("K")+
  scale_x_continuous(breaks = c(1:10))+
  theme_classic()
```
# Cross validation gets better up to 5 classes, then gets worse. Stuff after that is likely overbinning
```{r}
setwd("~/Desktop/petroica.preliminary/admixture/")
#read in input file in order to get list of input samples in order
sampling<-read.table("binary_fileset.fam")[,1]
sampling
#read in all ten runs and save each dataframe in a list
runs<-list()
#read in log files
for (i in 1:10){
  runs[[i]]<-read.table(paste0("binary_fileset.", i, ".Q"))
}

par(mfrow=c(1,1))
#plot each run
for (i in 1:5){
barplot(t(as.matrix(runs[[i]])), col=rainbow(i), ylab="Ancestry", border="black")
}
sampling
```

# treemix tree
```{r}
#run treemix on KU cluster using this code:
##!/bin/sh
##
##SBATCH --job-name=treemix              # Job Name
##SBATCH --nodes=1             # 40 nodes
##SBATCH --ntasks-per-node=1               # 40 CPU allocation per Task
##SBATCH --partition=sixhour            # Name of the Slurm partition used
##SBATCH --chdir=/home/d669d153/scratch/petroica/treemix    # Set working d$
##SBATCH --mem-per-cpu=2000           # memory requested
##SBATCH --time=360
#
##convert vcf into treemix file
#/home/d669d153/work/stacks-2.41/populations --in_vcf filtered.unlinked.85.vcf -O . --treemix -M pops.popmap.txt
##remove stacks header and standardize filename
#echo "$(tail -n +2 filtered.unlinked.85.p.treemix)" > unlinked.filtered.recode.p.treemix
##gzip file for input to treemix
#gzip unlinked.filtered.recode.p.treemix
#
##run treemix with m0
#/panfs/pfs.local/work/bi/bin/treemix-1.13/src/treemix -i unlinked.filtered.recode.p.treemix.gz -root aus -o treem0
#
##add 1 migration edge
#/panfs/pfs.local/work/bi/bin/treemix-1.13/src/treemix -i unlinked.filtered.recode.p.treemix.gz -m 1 -g treem0.vertices.gz #treem0.edges.gz -o treem1
#
##add 2 migration edges
#/panfs/pfs.local/work/bi/bin/treemix-1.13/src/treemix -i unlinked.filtered.recode.p.treemix.gz -m 1 -g treem1.vertices.gz #treem1.edges.gz -o treem2
#
##add 3 migration edges
#/panfs/pfs.local/work/bi/bin/treemix-1.13/src/treemix -i unlinked.filtered.recode.p.treemix.gz -m 1 -g treem2.vertices.gz #treem2.edges.gz -o treem3

#plot results
#source plotting functions that are distributed with treemix
source("~/Downloads/plotting_funcs.R")

#move into the treemix output directory and plot trees
setwd("~/Desktop/petroica.preliminary/treemix/")
#0 edge
plot_tree("treem0")
#1 edge
plot_tree("treem1", plus = 0.02, arrow=.1, ybar = 0, scale=F, lwd=1.5)
#2 edges
plot_tree("treem2")
#3 edges
plot_tree("treem3")


#plot to see how much variance is explained by each edge
#.994->m0
m=NULL
for(i in 0:3){
  m[i+1] <- get_f(paste0("treem",i))
}

m
plot(seq(0,3),m,pch="*",cex=2,col="blue", type="b",xlab="migration edge number", ylab="% explained variance")
```

# Dsuite
```{r}
#cluster commands
##copy in ruby scripts
#cp /home/d669d153/work/todi.subset.2022/dsuite/plot_d.rb .
#cp /home/d669d153/work/todi.subset.2022/dsuite/plot_f4ratio.rb .
##plot heatmap of D
#ruby plot_d.rb pops.popmap_unlinked_BBAA.txt plot_order.txt 0.7 unlinked_BBAA_D.svg
#
##plot heatmap of f4
#ruby plot_f4ratio.rb pops.popmap_unlinked_BBAA.txt plot_order.txt 0.2 unlinked_BBAA_f4ratio.svg
#
##run fbranch on the guidetree, with the Dtrios output ending in 'tree.txt' (guidetree must be rooted on #outgroup)
#/home/d669d153/work/Dsuite/Build/Dsuite Fbranch nwk.tre pops.popmap_unlinked_tree.txt > unlinked_Fbranch.txt
#
#module load python
##plot the fbranch heatmap with guidetree above
#/home/d669d153/work/Dsuite/utils/dtools.py unlinked_Fbranch.txt nwk.tre

#vis output
knitr::include_graphics("/Users/devder/Desktop/petroica.preliminary/dsuite/unlinked_BBAA_D.svg")
knitr::include_graphics("/Users/devder/Desktop/petroica.preliminary/dsuite/unlinked_BBAA_f4ratio.svg")
knitr::include_graphics("/Users/devder/Desktop/petroica.preliminary/dsuite/fbranch.svg")
```

